keps/sig-auth/5284-constrained-impersonation/README.md

```diff
@@ -0,0 +1,1445 @@
+verb `impersonate:user-info` with the related resources.
+
+#### Header `Impersonate-User` is set
+
+A subjectaccessreview
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentication.k8s.io
+    resource: users
+    name: someUser
+    verb: impersonate:user-info
+  user: impersonator
+```
+will be sent to the authorizer
+
+#### Header `Impersonate-Group` is set
+
+A subjectaccessreview
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentication.k8s.io
+    resource: groups
+    name: someGroup
+    verb: impersonate:user-info
+  user: impersonator
+```
+will be sent to the authorizer for each group.
+
+#### Header `Impersonate-Uid` is set
+
+A subjectaccessreview
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentication.k8s.io
+    resource: uids
+    name: someUID
+    verb: impersonate:user-info
+  user: impersonator
+```
+will be sent to the authorizer.
+
+#### Header with prefix `Impersonate-Extra-` is set
+
+A subjectaccessreview
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentication.k8s.io
+    resource: userextras
+    subresource: extraKey
+    name: extraValue
+    verb: impersonate:user-info
+  user: impersonator
+```
+will be sent to the authorizer for each key and value pair.
+
+### Verb `impersonate:serviceaccount`
+
+Same as legacy impersonation, when the request has the header of `Impersonate-User`, and the value of the header
+has a prefix of `system:serviceaccount:`, apiserver will check verb `impersonate:serviceaccount` with the authorizer
+using the subjectaccessreview:
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentication.k8s.io
+    resource: serviceaccounts
+    name: serviceaccount-name
+    namespace: serviceaccount-namespace
+    verb: impersonate:serviceaccount
+  user: impersonator
+```
+
+### Verb `impersonate:node` and `impersonate:scheduled-node`
+
+When the request has the header of `Impersonate-User`, and the value has the prefix of `system:nodes`,
+this verb is checked instead of `impersonate:user-info`. The subjectaccessreview below will be sent
+to the authorizer:
+```yaml
+apiVersion: authorization.k8s.io/v1
+kind: SubjectAccessReview
+spec:
+  resourceAttributes:
+    group: authentication.k8s.io
+    resource: nodes
+    name: someNode
+    verb: impersonate:nodes
+  user: impersonator
+```
+
+`impersonate:scheduled-node` is a special verb to check when two conditions are met:
+1. The impersonator is impersonating a node by setting the header `Impersonate-User` and the value has a prefix
+of `system:node:` on the request.
+2. The user info of the impersonator has an extra with key `authentication.kubernetes.io/node-name`, and the value
```

Comment:
Is the impersonator required to be a service account, or just have the matching `authentication.kubernetes.io/node-name` extra info value?